lightning/src/ln/channelmanager.rs

```diff
@@ -12477,6 +12477,23 @@ where
 				);
 
 				if self.default_configuration.manually_handle_bolt12_invoices {
+					// Prevent duplicate `InvoiceReceived` events by checking if this invoice was already processed.
+					let already_present = self.pending_events.lock().unwrap().iter().any(|(e, _)| {
+						matches!(e, Event::InvoiceReceived { payment_id: existing_id, .. } if existing_id == &payment_id)
+					});
```

Comment:
Can't we do a lookup in `PendingOutboundPayments` by payment id and check it's state? We could transition it to the `InvoiceReceived` state before generating the event.